broker/lib/fabrik/DirectorManager.js

```diff
@@ -177,7 +189,179 @@ class DirectorManager extends BaseManager {
+    const scheduled = _.get(params, 'scheduled') || false;
+    const dbUpdate = _.get(params, '_serviceFabrikDbUpdate') || false;
+    const policyApplicable = config.enable_bosh_rate_limit;
+    _.omit(params, 'scheduled');
+    _.omit(params, '_serviceFabrikDbUpdate');
+
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false,
+      'dbUpdate': false,
+      'policyApplicable': config.enable_bosh_rate_limit,
+      'directorError': false
+    };
+
+    return this.executePolicy(scheduled, action, deploymentName, dbUpdate, policyApplicable)
+      .then(checkResults => {
+        if (checkResults.directorError) {
+          decisionMaker.directorError = true;
+          return Promise.resolve(true);
+        }
+        const shouldRunNow = checkResults.shouldRunNow;
+
+        if (!decisionMaker.policyApplicable) {
+          return Promise.resolve();
+        }
+        if (dbUpdate) {
+          decisionMaker.shouldRunNow = true;
+          decisionMaker.dbUpdate = true;
+          return Promise.resolve();
+        }
+        if (shouldRunNow) {
+          decisionMaker.shouldRunNow = shouldRunNow;
+          return boshOperationCache.containsDeployment(deploymentName).then(inCache => {
+            if (inCache) {
+              decisionMaker.cached = true;
+            }
+            return Promise.resolve();
+          });
+        }
+        if (!scheduled) {
+          return Promise.resolve(true);
+        }
+        throw new DeploymentDelayed(deploymentName);
+      })
+      .then((storeInCache) => {
+        if (storeInCache) {
+          // stagger here by putting it into etcd cache and return promise
+          return boshOperationCache.store(this.plan.id, deploymentName, params, args)
+            .then(() => {
+              decisionMaker.cached = true;
+              return Promise.resolve();
+            })
+            .catch(err => {
+              logger.error('Error updating etcd cache', err);
+              throw new CacheUpdateError(deploymentName);
+            });
+        }
+      })
+      .then(() => {
+        if (decisionMaker.shouldRunNow && decisionMaker.cached) {
+          //the deployment was cached in etcd earlier; remove it from cache and proceed
+          return boshOperationCache.deleteDeploymentFromCache(deploymentName);
+        }
+      })
+      .then(() => {
+        if (decisionMaker.directorError) {
+          throw new DeploymentDelayed(deploymentName);
+        } else if (decisionMaker.dbUpdate || decisionMaker.shouldRunNow || !decisionMaker.policyApplicable) {
+          return;
+        } else {
+          throw new DeploymentDelayed(deploymentName);
+        }
+      })
+      .then(() => this._createOrUpdateDeployment(deploymentName, params, args, scheduled))
+      .then(taskId => {
+        const out = {
+          'cached': false
+        };
+        if (decisionMaker.cached) {
```

Comment:
when will this be true? if cached wont you throw the exception?see LN 324 - that should ensure no entries exist in etcd if its to run now?